### PR TITLE
[site] Add dynamic section table

### PR DESCRIPTION
**Pull request by towoe**
Shortcode for inserting a generated table of content for the current
folder.
This can be used in Markdown files in order to have an always up-to-date
list of the content of the folder.

Supports a "type" parameter which can be set to "section" in order to
list only subsection. Will ignore other content such as normal pages.
Parameter "depth" controls the level to which the content is listed.
Defaults to two if not set.

Fixes lowrisc/opentitan#1419

Signed-off-by: Tobias Wölfel <tobias.woelfel@mailbox.org>

### CODE REVIEW

File: site/docs/layouts/shortcodes/sectionContent.html [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                        <a href="{{.RelPermalink}}">{{- .LinkTitle }}</a>
+                    {{- else }}
+                        {{- replace .File.Dir .Parent.File.Dir "" | replaceRE "/" "-" | replaceRE "-$" ":" }}
+                        <a href="{{.RelPermalink}}">{{- .File.TranslationBaseName }}</a>
+                    {{- end}}
+                {{- end}}
+            {{- end}}
+            {{- if lt $depthCnt $depthLimit }}
+                <ul class="
+                    ">
```
Comment (gkelly) [inline site/docs/layouts/shortcodes/sectionContent.html:40]: This seems like there's a missing class name here or a broken copy/paste?

Comment (towoe) [reply site/docs/layouts/shortcodes/sectionContent.html]: Thanks @gkelly for spotting that. It was a testing relic.

File: site/docs/assets/scss/_sectioncontent.scss [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.sectioncontent {
+
```
Comment (gkelly) [inline site/docs/assets/scss/_sectioncontent.scss:2]: nit: Drop newline.

File: site/docs/layouts/shortcodes/sectionContent.html [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    {{- $startSection := .startSection }}
+    {{- $currentSection := .currentSection }}
+    {{- $contentType := .contentType }}
+    {{- $depthLimit := .depthLimit }}
+    {{- $depthCnt := add .depthCnt 1 }}
+
+    {{- with .startSection }}
+        <li>
+            {{- with $currentSection }}
+                {{- if (or (not (in $contentType "section")) .IsSection) }}
```
Comment (gkelly) [inline site/docs/layouts/shortcodes/sectionContent.html:29]: Can you use `(eq $contentType "section")` here? The `in` invocation seems wasteful.

File: site/docs/layouts/shortcodes/sectionContent.html [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                    {{- if .LinkTitle }}
+                        <a href="{{.RelPermalink}}">{{- .LinkTitle }}</a>
+                    {{- else }}
+                        {{- replace .File.Dir .Parent.File.Dir "" | replaceRE "/" "-" | replaceRE "-$" ":" }}
+                        <a href="{{.RelPermalink}}">{{- .File.TranslationBaseName }}</a>
+                    {{- end}}
+                {{- end}}
+            {{- end}}
```
Comment (gkelly) [inline site/docs/layouts/shortcodes/sectionContent.html:37]: nit: Un-cuddle `end}}` -> `end }}`.

File: site/docs/assets/scss/_sectioncontent.scss (hunk 2 above)
Comment (towoe) [reply site/docs/assets/scss/_sectioncontent.scss]: done

File: site/docs/layouts/shortcodes/sectionContent.html (hunk 4 above)
Comment (towoe) [reply site/docs/layouts/shortcodes/sectionContent.html]: done

File: site/docs/layouts/shortcodes/sectionContent.html (hunk 3 above)
Comment (towoe) [reply site/docs/layouts/shortcodes/sectionContent.html]: That is indeed possible. Thanks for the tip.

**Comment by towoe** [comment]
Thanks a lot for the review @gkelly

**Comment by gkelly** [comment]
Looking at the rendered version: [General Documentation _ OpenTitan Documentation.pdf](https://github.com/lowRISC/opentitan/files/4719740/General.Documentation._.OpenTitan.Documentation.pdf), there's an entry that looks like `install_instructions: index`. Can that just be the `install_instructions` linkified? If it has a `title:` in the frontmatter does it look better?

**Comment by towoe** [comment]
> Looking at the rendered version: [General Documentation _ OpenTitan Documentation.pdf](https://github.com/lowRISC/opentitan/files/4719740/General.Documentation._.OpenTitan.Documentation.pdf), there's an entry that looks like `install_instructions: index`. Can that just be the `install_instructions` linkified? If it has a `title:` in the frontmatter does it look better?

I did format this on purpose because I thought it looked better especially for the sw folder.
But this was just my personal preference. Happy to change it.




I think using only the directory path will result in some other corner cases where it will not work (README.md in sw/).
I did not want to put to much effort into this as you already mentioned, it will look different for files where `title:` is set in the frontmatter. I only added this as a backup so we have some text for the link. From my side the goal should be that "real" documentation should use the frontmatter, with this approach we can gradually get there.

**Comment by imphil** [comment]
Looks good, leaving the fine details to @gkelly.

**Comment by gkelly** [comment]
> > Looking at the rendered version: [General Documentation _ OpenTitan Documentation.pdf](https://github.com/lowRISC/opentitan/files/4719740/General.Documentation._.OpenTitan.Documentation.pdf), there's an entry that looks like `install_instructions: index`. Can that just be the `install_instructions` linkified? If it has a `title:` in the frontmatter does it look better?
> 
> I did format this on purpose because I thought it looked better especially for the sw folder.
> But this was just my personal preference. Happy to change it.
> 
> 
> 
> 
> I think using only the directory path will result in some other corner cases where it will not work (README.md in sw/).
> I did not want to put to much effort into this as you already mentioned, it will look different for files where `title:` is set in the frontmatter. I only added this as a backup so we have some text for the link. From my side the goal should be that "real" documentation should use the frontmatter, with this approach we can gradually get there.

Got it, thanks for the explanation. In that case this is good to go! Thanks for making this much nicer to use.

**Comment by imphil** [comment]
@towoe this is your chance to merge your first own pull request! I'm leaving it open for you.

**Comment by towoe** [comment]
Thanks @imphil ;) and thanks @gkelly  for the review.

**Pull request merged**